pennylane/qnode.py

```diff
@@ -596,6 +716,81 @@ def check_op(op):
         ret = self.device.execute(self.queue, self.ev, self.variable_ops)
         return self.output_conversion(ret)
 
+    def metric_tensor(self, *args, **kwargs):
+        """Evaluate the value of the metric tensor.
+
+        Args:
+            args : qfunc positional arguments
+            kwargs : qfunc keyword arguments
+
+        Keyword Args:
+            diag_approx (bool): If ``True``, forces the diagonal
+                approximation. Default is ``False``.
+
+        Returns:
+            array[float]: measured values
+        """
+        diag_approx = kwargs.pop("diag_approx", False)
+
+        if not self.ops or not self.cache:
+            # construct the circuit
+            self.construct(args, kwargs)
+
+        # temporarily store keyword arguments
+        keyword_values = {}
+        keyword_values.update({k: np.array(list(_flatten(v))) for k, v in self.keyword_defaults.items()})
+        keyword_values.update({k: np.array(list(_flatten(v))) for k, v in kwargs.items()})
+
+        # temporarily store the free parameter values in the Variable class
+        Variable.free_param_values = np.array(list(_flatten(args)))
+        Variable.kwarg_values = keyword_values
+
+        if not self._metric_tensor_subcircuits:
+            self.construct_metric_tensor(args, diag_approx=diag_approx, **kwargs)
+
+        tensor = np.zeros([self.num_variables, self.num_variables])
+
+        # execute any constructed metric tensor subcircuits
+        for params, circuit in self._metric_tensor_subcircuits.items():
+            self.device.reset()
+
+            s = np.array(circuit['scale'])
+            V = circuit['eigenbasis']
+
+            if not diag_approx:
+                # block diagonal approximation
+
+                unitary_op = pennylane.ops.QubitUnitary(V, wires=list(range(self.num_wires)), do_queue=False)
+                self.device.execute(circuit['queue'] + [unitary_op], circuit['observable'])
+                probs = list(self.device.probability().values())
+
+                first_order_ev = np.zeros([len(params)])
+                second_order_ev = np.zeros([len(params), len(params)])
+
+                for idx, Lambda in circuit['first_order']:
+                    first_order_ev[idx] = Lambda @ probs
+
+                for idx, Lambda in circuit['second_order']:
+                    second_order_ev[idx] = Lambda @ probs
+                    second_order_ev[idx[1], idx[0]] = second_order_ev[idx]
```

Comment:
Oh, this is because <psi|K_i K_j|psi> is computed in the line above, and since we assume K_i, K_j commute, we avoid recalculating <psi|K_j K_i|psi>. Essentially, the `second_order_ev` matrix is just symmetric.

Reply:
Done